src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java

```diff
@@ -109,11 +109,25 @@ public void execute(Run run, FlowNode flowNode) {
         // Every found flow node of the DAG is added to the BuildPipeline instance.
         scanner.forEach(pipeline::add);
 
-        final SpanContext spanContext = tracer.extract(Format.Builtin.TEXT_MAP, new BuildTextMapAdapter(buildSpanAction.getBuildSpanPropatation()));
+        final TraceSpan.TraceSpanContext traceSpanContext = buildSpanAction.getBuildSpanContext();
         final BuildPipelineNode root = pipeline.buildTree();
 
+        final List<PayloadMessage> spanBuffer = new ArrayList<>();
+        collectTraces(run, spanBuffer, buildData, root, traceSpanContext);
+
         try {
-            sendTrace(tracer, run, buildData, root, spanContext);
+            if(!spanBuffer.isEmpty()) {
+                final List<PayloadMessage> spanSendBuffer = new ArrayList<>(SIZE_SPANS_SEND_BUFFER);
+                for(int i = 0; i < spanBuffer.size(); i++) {
+                    spanSendBuffer.add(spanBuffer.get(i));
+
+                    // Send every 100 spans or the last one.
+                    if(spanSendBuffer.size() == SIZE_SPANS_SEND_BUFFER || i == (spanBuffer.size() - 1)) {
+                        agentHttpClient.send(Collections.unmodifiableList(spanSendBuffer));
+                        spanSendBuffer.clear();
+                    }
+                }
+            }
```

Comment:
Avoid unbounded payloads

Reply:
Could this be an internal of the client?

Reply:
Yes, totally. 👌🏼